infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -187,220 +225,328 @@ private boolean canHandleReconfigurations() {
+     * Failure/Healing detector task. Detects faults in the cluster and heals servers also corrects wrong epochs:
+     *  - correct wrong epochs by resealing the servers and updating trailing layout servers.
+     *    Fetch quorum layout and update the epoch according to the quorum.
+     *  - check if current layout slot is unfilled then update layout to latest one.
+     *  - Looking for link failures in the cluster, handle failure if found
+     *  - bootstrap sequencer if needed
+     *  - handle healed nodes
+     * </pre>
+     *
+     * @param pollReport cluster status
+     * @return async detection task
      */
-    private void runHealingDetectorTask() {
+    private CompletableFuture<DetectorTask> runFailureDetectorTask(PollReport pollReport) {
 
-        if (!healingDetectorFuture.isDone()) {
-            log.debug("Cannot initiate new healing polling task. Polling in progress.");
-            return;
+        if (!pollReport.getClusterState().isReady()) {
+            log.info("Cluster state is not ready: {}", pollReport.getClusterState());
+            return DETECTOR_TASK_SKIPPED;
         }
 
-        healingDetectorFuture = detectionTaskWorkers.submit(() -> {
+        return CompletableFuture.supplyAsync(() -> {
 
-            CorfuRuntime corfuRuntime = getCorfuRuntime();
-            final Layout layout = serverContext.copyManagementLayout();
-            final PollReport pollReport = healingDetector.poll(layout, corfuRuntime);
+            // Corrects out of phase epoch issues if present in the report. This method
+            // performs re-sealing of all nodes if required and catchup of a layout server to
+            // the current state.
+            correctWrongEpochs(pollReport);
 
-            clusterStateContext.updateResponsiveNodes(pollReport.getChangedNodes());
-            pollReport.getClusterStateMap().forEach((s, clusterStatus) ->
-                    clusterStateContext.updateNodeState(clusterStatus, layout));
+            try {
+                if (pollReport.isCurrentLayoutSlotUnFilled()) {
+                    handleFailure(Collections.emptySet(), pollReport).get();
+                    return DetectorTask.COMPLETED;
+                }
 
-            Set<String> healedNodes = new TreeSet<>(recommendationEngine
-                    .healedServers(clusterStateContext.getClusterState(), layout));
+                if (!pollReport.getWrongEpochs().isEmpty()) {
+                    log.debug("Wait for next iteration. Poll report contains wrong epochs: {}",
+                            pollReport.getWrongEpochs()
+                    );
+                    return DetectorTask.COMPLETED;
+                }
+            } catch (Exception ex) {
+                log.error("Can't fill slot. Poll report: {}", pollReport, ex);
+            }
 
-            if (healedNodes.isEmpty()) {
-                return;
+            DetectorTask healing = handleHealing(pollReport, serverContext.copyManagementLayout());
```

Comment:
Weren't we planning to run healing and failure handling in parallel?
What happens in case a failure handling is required before healing can take place?
Can you add a comment explaining that scenario?
The java doc in the method states failure handling first and then healing handling.

Reply:
1. If the list of nodes to heal is empty then nothing to heal.
2. I believe there are no cases when it's required by design - fail node then heal the current node. Also if the node can't heal itself it doesn't prevent the node to go to handleFailure step
3. javadoc - fixed

I also added some comments